src/ram/src/ram.cpp

```diff
@@ -0,0 +1,609 @@
+//   this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+///////////////////////////////////////////////////////////////////////////////
+
+#include "ram/ram.h"
+
+#include "db_sta/dbNetwork.hh"
+#include "layout.h"
+#include "sta/FuncExpr.hh"
+#include "sta/Liberty.hh"
+#include "sta/PortDirection.hh"
+#include "utl/Logger.h"
+
+namespace ram {
+
+using odb::dbBlock;
+using odb::dbBPin;
+using odb::dbBTerm;
+using odb::dbInst;
+using odb::dbMaster;
+using odb::dbNet;
+using odb::dbRow;
+
+using utl::RAM;
+
+using std::array;
+using std::vector;
+
+////////////////////////////////////////////////////////////////
+
+RamGen::RamGen() : db_(nullptr), logger_(nullptr)
+{
+}
+
+void RamGen::init(odb::dbDatabase* db, sta::dbNetwork* network, Logger* logger)
+{
+  db_ = db;
+  network_ = network;
+  logger_ = logger;
+}
+
+dbInst* RamGen::makeCellInst(
+    Cell* cell,
+    const std::string& prefix,
+    const std::string& name,
+    dbMaster* master,
+    const vector<std::pair<std::string, dbNet*>>& connections)
+{
+  const auto inst_name = fmt::format("{}.{}", prefix, name);
+  auto inst = dbInst::create(block_, master, inst_name.c_str());
+  for (auto& [mterm_name, net] : connections) {
+    auto mterm = master->findMTerm(mterm_name.c_str());
+    if (!mterm) {
+      logger_->error(
+          RAM, 11, "term {} of cell {} not found.", name, master->getName());
+    }
+    auto iterm = inst->getITerm(mterm);
+    iterm->connect(net);
+  }
+
+  cell->addInst(inst);
+  return inst;
+}
+
+dbNet* RamGen::makeNet(const std::string& prefix, const std::string& name)
+{
+  const auto net_name = fmt::format("{}.{}", prefix, name);
+  return dbNet::create(block_, net_name.c_str());
+}
+
+dbBTerm* RamGen::makeBTerm(const std::string& name)
+{
+  auto net = dbNet::create(block_, name.c_str());
+  auto bTerm = dbBTerm::create(net, name.c_str());
+
+  return bTerm;
+}
+
+dbBTerm* RamGen::makeOutputBTerm(const std::string& name)
+{
+  auto net = dbNet::create(block_, name.c_str());
+  auto bTerm = dbBTerm::create(net, name.c_str());
+
+  bTerm->setIoType(odb::dbIoType::OUTPUT);
+  return bTerm;
+}
```

Comment:
It probably makes more sense to have just one `makeBTerm` function and then pass in the IoType as an argument. Less code that way.